[Cutlass] Implement cutlass epilogue visitor python codegen

mlazos opened the pull request:
This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150910
* #152390
* #150909
* #150908
* #150907
* #151406
* #150906
* #151713
* #151405
* __->__ #150905
* #152306
* #152305

Previously merged:
* #150904
* #150903
* #150346
* #150345
* #150344

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

mlazos committed:
[Cutlass] Implement cutlass epilogue visitor python codegen

[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"


[Cutlass] Updates to evt codegen

updates to python codegen

More python codegen updates

More python codegen updates

updates to evt python

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only support add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only support add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only support add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented:
tests failing ?

mlazos commented:
> tests failing ?

I don't think this these are related believe it or not, but I'll rebase.
Main is was very red.

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,201 @@
+            ),
+        ):
+            for node in epilogue_nodes:
+                assert isinstance(node, ComputedBuffer)
```
Where do we handle upcasting the mm output to fp32 ? / making sure epilogue is done in fp32 ?

mlazos replied:
I was able to compile mixed dtype epilogues so it will respect whatever dtype the epilogue nodes have.

eellison commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,201 @@
+        """
+        self.accumulator_node_name: str = accumulator_node_name  #
+        self.output: IndentedBuffer = IndentedBuffer(1)  # The output buffer for codegen
+        self.var_counter: int = (
```
nit: `itertools.count`

eellison commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,201 @@
+                data = node.data
+                # TODO mlazos: relax this, cutlass supports reductions and other ops
+                assert isinstance(data, Pointwise)
+                index = data._index(data.ranges)
+                result = data.inner_fn(index)
```
Can you say a bit more about how index is handled ? As far as I know, we are not actually doing any indexing in the cutlass generated code.

mlazos replied:
So now I handled indexing on reads, we check that the indexing strides match the strides of the read's layout. The only remaining question I have is is this needed for writes? I assumed not because writes always use the indexer of the layout of the buffer being written to.

eellison commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,201 @@
+            self.reads.add(name)
+            return name
+
+    def store(
+        self, name: Any, index: Any = None, value: Any = None, mode: Any = None
+    ) -> None:
+        self.writes.add(name)
+        return None
```
Similarly, surprising to me that we are not using index, or value here..

eellison reviewed (commented):
Looking good ! A few questions:

- how are we handling different indexing ? Since we are ignoring it currently, is it sufficient to check that the various indexes are broadcasts of each other ? Since we are codegening broadcasts ?

- Do we ever have multiple outputs ? How do we handle the epilogue fusion potentially still outputting or not outputting the original mm output ? Within triton, we check `V.graph.removed_buffers` and ignore writes to removed buffers.

Could we add tests for 
```
def foo():
 .... 
    out = mm(...)
    return out, out.relu()
```
Versus:
```
def foo():
 .... 
    out = mm(...)
    return out.relu()
```

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,201 @@
+            if name in ("load", "store"):
+                return line
+            else:
+                var = self.parent_handler._tmp_var()
+                self.parent_handler.output.writeline(f"{var} = {line}")
+            return var
```
We're basically doing a mix of what Ops does and what kernel rendering does, which is why I need this. In the `_default` impl, if the op isn't supported in the parent handler we throw `NotImplemented` with the current way the `_ComputeAssignmentFormatter` is setup.

mlazos replied:
I cleaned this up quite a bit so hopefully this is clearer now - basically the reason there are two classes is that the `_default` impl of the `_AssignmentFormatter` will automatically throw `NotImplemented` for unsupported ops. The reason to do this is that it separates the getattr rerouting away from the class which has a lot of other logic in it (`CutlassEVTCodegen`). This is analogous to the relationship between `Kernel`(the ops handler) and `CSE`(var naming/assignment) in normal triton codegen.

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


Previously merged:
* #150904
* #150903
* #150346
* #150345
* #150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


Previously merged:
* #150904
* #150903
* #150346
* #150345
* #150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos commented:
> Looking good ! A few questions:
> 
> * how are we handling different indexing ? Since we are ignoring it currently, is it sufficient to check that the various indexes are broadcasts of each other ? Since we are codegening broadcasts ?
> * Do we ever have multiple outputs ? How do we handle the epilogue fusion potentially still outputting or not outputting the original mm output ? Within triton, we check `V.graph.removed_buffers` and ignore writes to removed buffers.
> 
> Could we add tests for
> 
> ```
> def foo():
>  .... 
>     out = mm(...)
>     return out, out.relu()
> ```
> 
> Versus:
> 
> ```
> def foo():
>  .... 
>     out = mm(...)
>     return out.relu()
> ```

So now I aggregate all of the last usages of the constituent scheduler nodes and on each store check if the name is in the last usages of the epilogue, if so I ignore. This should be equivalent, I just wasn't sure of the semantics of removed_buffers.

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


Previously merged:
* #150904
* #150903
* #150346
* #150345
* #150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,240 @@
+    def store(
+        self, name: Any, index: Any = None, value: Any = None, mode: Any = None
+    ) -> None:
+        if name not in self.last_usages:
+            value_to_write = value
+            if not self.store_name_to_value:
+                # EVT requires an output to be named D lol
+                # so rename the first store to D
```
Do we need to check `_check_indexing` here as well ?

mlazos replied:
I don't think it's possible to have different indexing on stores because ComputedBuffers always grab the indexer from its layout, there isn't a way to intercept the indexing to change it like would happen with a PermuteView. TBH I'll add it just to be safe.

mlazos commented on torch/_inductor/codegen/cuda/cutlass_python_evt.py:
```diff
@@ -0,0 +1,240 @@
+        self.accumulator_node_name: str = accumulator_node_name  #
+        self.output: IndentedBuffer = IndentedBuffer(1)  # The output buffer for codegen
+        self.var_counter: Iterator[int] = itertools.count()
+        self.store_name_to_value: dict[str, OpsValue] = (
```
lintrunner doesn't think so lol (it's because of the comment making the line long)

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


Previously merged:
* #150904
* #150903
* #150346
* #150345
* #150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

mlazos committed:
Update on "[Cutlass] Implement cutlass epilogue visitor python codegen"

This PR implements the second codegen task of CUTLASS EVT: translating inductor epilogue nodes into python code that will be traced by the EVT infra. 

Details:
The implementation uses a simple ops wrapper which only supports add and mul pointwise ops today (to be extended in the future). This ops wrapper generates python code from inner_fn of the epilogue nodes in the format EVT expects. The main caveat is that one of the outputs needs to be named "D" and the accumulator input needs to be named "acc". Reads/writes are named according to the inductor buffer names otherwise. 


Previously merged:
* #150904
* #150903
* #150346
* #150345
* #150344

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
